src/content/mods.cpp

```diff
@@ -360,51 +360,127 @@ void ModConfiguration::checkConflictsAndDeps()
-	std::set<std::string> modnames;
-	for (const ModSpec &mod : m_unsatisfied_mods) {
-		modnames.insert(mod.name);
+	PrioritySortedMod(ModSpec *mod) : spec(mod) {}
+
+	static bool sorter(const PrioritySortedMod *a, const PrioritySortedMod *b)
+	{
+		// Return true means: move 'a' towards begin() in std::sort
+		return a->deps_chain.size() > b->deps_chain.size();
+	}
+
+	bool addDependingMod(const std::string &modname,
+			std::unordered_map<std::string, PrioritySortedMod> &mods_by_name);
+
+	std::unordered_set<std::string> deps_chain;
+	ModSpec *spec;
+};
+
+bool PrioritySortedMod::addDependingMod(const std::string &modname,
+		std::unordered_map<std::string, PrioritySortedMod> &mods_by_name)
+{
+	if (deps_chain.find(modname) != deps_chain.end()) {
+		if (modname == spec->name) {
+			// Step 4: Reached the very same mod again.
+			// Stop the circular dependencies
+
+			warningstream << "Detected circular dependencies in mod '" <<
+					modname << "'" << std::endl;
+			return false;
+		}
+		return true; // The dependency trees merge. Nothing unusual
 	}
 
-	// Step 2: get dependencies (including optional dependencies)
-	// of each mod, split mods into satisfied and unsatisfied
-	std::list<ModSpec> satisfied;
-	std::list<ModSpec> unsatisfied;
-	for (ModSpec mod : m_unsatisfied_mods) {
-		mod.unsatisfied_depends = mod.depends;
-		// check which optional dependencies actually exist
-		for (const std::string &optdep : mod.optdepends) {
-			if (modnames.count(optdep) != 0)
-				mod.unsatisfied_depends.insert(optdep);
+	deps_chain.emplace(modname);
+
+	const auto &map_end = mods_by_name.end();
+
+	// Step 3: Follow the dependeny path and "push up" each
+	// mod which is required by 'modname'
+	for (const std::string &dependency : spec->depends) {
+		const auto &mod = mods_by_name.find(dependency);
+		if (mod == map_end) {
+			// Hard dependency not found
+			if (modname == spec->name)
+				spec->unsatisfied_depends.emplace(dependency);
+			return false;
 		}
-		// if a mod has no depends it is initially satisfied
-		if (mod.unsatisfied_depends.empty())
-			satisfied.push_back(mod);
-		else
-			unsatisfied.push_back(mod);
+
+		// Attempt to add "modname" to the depending mods
+		if (!mod->second.addDependingMod(modname, mods_by_name))
+			return false;
+	}
+
+	for (const std::string &dependency : spec->optdepends) {
+		const auto &mod = mods_by_name.find(dependency);
+		// It does not matter whether optional dependencies can be added,
+		// hence ignore the return value.
+		if (mod != map_end)
+			mod->second.addDependingMod(modname, mods_by_name);
+	}
+	return true;
+}
+
+void ModConfiguration::resolveDependencies()
+{
+	std::unordered_map<std::string, PrioritySortedMod> mods_by_name;
+	for (ModSpec &mod : m_unsatisfied_mods) {
+		// Reduce memory copy time by passing 'mod' as pointer
+		// 'm_unsatisfied_mods' owns the value!
+		mods_by_name.emplace(std::pair<std::string, PrioritySortedMod>(
+				mod.name, &mod));
```

Comment:
```suggestion
		mods_by_name[mod.name] = PrioritySortedMod(&mod);
```

Reply:
Indexing maps does implicitly call the underlying `PrioritySortedMod` constructor with zero arguments. Worked around this issue by using the default value `nullptr`.

Reply:
Ahhh, right. This should work then even without a default constructor:
```diff
-		mods_by_name.emplace(std::pair<std::string, PrioritySortedMod>(
-				mod.name, &mod));
+		mods_by_name.emplace(mod.name, &mod);
```

If not that's also okay, doesn't really matter.